### PR TITLE
finished spawn system

### PR DESCRIPTION
**Pull request by JayDeeThomas**

### CODE REVIEW

File: Assets/Scripts/Spawn System/PlayerRespawn.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                }
+            else if (allowBacktrack) {
+                respawnPoint = newSpawn.postion;
+                }
+            }
+        }
+
+    //spawn character in
+    public void respawnPlayer() {
+        player.transform.position = respawnPoint;
```
Comment (smlendzian) [inline Assets/Scripts/Spawn System/PlayerRespawn.cs:47]: Can replace this line with: this.transform.position = respawnPoint;

File: Assets/Scripts/Spawn System/PlayerRespawn.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+
+//attach to player
+public class PlayerRespawn : MonoBehaviour
+{
+    GameObject player;
```
Comment (smlendzian) [inline Assets/Scripts/Spawn System/PlayerRespawn.cs:10]: Can remove player object (see comment below where the object can be referenced using this)

File: Assets/Scripts/Spawn System/PlayerRespawn.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    public Vector3 respawnPoint;
+    //allows reuse of old checkpoints
+    public bool allowBacktrack = false;
+
+
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        player = this.gameObject;
```
Comment (smlendzian) [inline Assets/Scripts/Spawn System/PlayerRespawn.cs:21]: can remove player object (see below)

File: Assets/Scripts/Spawn System/Checkpoint.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Checkpoint : MonoBehaviour
+{
+    
+    public bool activated = false; //stops player from reactivation old check points
+    public Vector3 postion; //spawn location
```
Comment (smlendzian) [inline Assets/Scripts/Spawn System/Checkpoint.cs:9]: these should be private
You can create a private serialized field using the syntax [SerializeField] private bool activated

Comment (smlendzian) [inline Assets/Scripts/Spawn System/Checkpoint.cs:9]: These should be properties

File: Assets/Scripts/Spawn System/SpawnPoint.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        // places player at level spawn at start of level 
+          //2nd load onwards
+        player.respawnPoint = position;
+        player.respawnPlayer();
+        
+        
+    }
+
+    
+}
```
Comment (cadenceblorbo) [inline Assets/Scripts/Spawn System/SpawnPoint.cs:29]: Probably best to remove this class entirely, and justhave an option to pass an initial checkpoint object into the playerrespawn through a serializefield. I don't think we need the non-revisiting property of this behavior (to my knowledge, the only thing that would functionally change), but if we want that option, adding a toggle for non-revisiting to checkpoint is probably the best way to go about it.

File: Assets/Scripts/Spawn System/Checkpoint.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    [SerializeField] int height = 1; //y position to spawn player at
```
Comment (cadenceblorbo) [inline Assets/Scripts/Spawn System/Checkpoint.cs:10]: this should be a Vector3 offset to allow for more dynamic behavior

File: Assets/Scripts/Spawn System/Checkpoint.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        postion = new Vector3(this.transform.position.x, height, this.transform.position.z);
```
Comment (cadenceblorbo) [inline Assets/Scripts/Spawn System/Checkpoint.cs:15]: A better implementation would be to do `position = this.transform.position + offset;`, currently the y coordinate is being set in world space which will raise difficulties if the checkpoint isn't at y=0.

File: Assets/Scripts/Spawn System/PlayerRespawn.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+
+//attach to player
+public class PlayerRespawn : MonoBehaviour
+{
+    GameObject player;
+    SpawnPoint spawn;
```
Comment (cadenceblorbo) [inline Assets/Scripts/Spawn System/PlayerRespawn.cs:11]: can change this to a serializefield containing a checkpoint (see recommended changes in SpawnPoint)

File: Assets/Scripts/Spawn System/PlayerRespawn.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    public Vector3 respawnPoint;
+    //allows reuse of old checkpoints
+    public bool allowBacktrack = false;
+
+
+
+    // Start is called before the first frame update
+    void Start()
```
Comment (cadenceblorbo) [inline Assets/Scripts/Spawn System/PlayerRespawn.cs:19]: update to non-spawn point using logic (see SpawnPoint for details)

File: Assets/Scripts/Spawn System/PlayerRespawn.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        //put player at leevel spawn to start 1st load of player
+        
+        respawnPoint = spawn.position;
+        respawnPlayer();
+    }
+
+   
+    private void OnTriggerEnter(Collider other) {
+        //change respawn to current spawn
+        if (other.gameObject.CompareTag("Respawn")) {
```
Comment (cadenceblorbo) [inline Assets/Scripts/Spawn System/PlayerRespawn.cs:32]: change to 
```
if (!other.gameObject.CompareTag("Respawn")) {
    return;
}
``` to reduce nesting

**Review by cadenceblorbo — CHANGES_REQUESTED** [review]
See comments for recommended changes to system

File: Assets/Scripts/Spawn System/SpawnPoint.cs (hunk 5 above)
Comment (cadenceblorbo) [reply Assets/Scripts/Spawn System/SpawnPoint.cs:29]: (this should also make loading saves easier)

**Pull request merged**